src/SAML2/XML/saml/AuthnContext.php

```diff
@@ -0,0 +1,266 @@
+     */
+    public static function fromXML(DOMElement $xml): object
+    {
+        /** @var \DOMElement[] $authnContextClassRef */
+        $authnContextClassRef = Utils::xpQuery($xml, './saml_assertion:AuthnContextClassRef');
+        Assert::maxCount($authnContextClassRef, 1);
+
+        /** @var \DOMElement[] $authnContextDeclRef */
+        $authnContextDeclRef = Utils::xpQuery($xml, './saml_assertion:AuthnContextDeclRef');
+        Assert::maxCount($authnContextDeclRef, 1);
+
+        /** @var \DOMElement[] $authnContextDecl */
+        $authnContextDecl = Utils::xpQuery($xml, './saml_assertion:AuthnContextDecl');
+        Assert::maxCount($authnContextDecl, 1);
+
+        if (!empty($authnContextClassRef)) {
+            Assert::oneOf(
+                [],
+                [$authnContextDecl, $authnContextDeclRef],
+                'Can only have one of AuthnContextDecl/AuthnContextDeclRef'
+            );
+        } else {
+            Assert::false(
+                empty($authnContextDecl) && empty($authnContextDeclRef),
+                'You need either an AuthnContextDecl or an AuthnContextDeclRef'
+            );
+        }
+
+        /** @var \DOMElement[] $authenticatingAuthorities */
+        $authenticatingAuthorities = Utils::xpQuery($xml, './saml_assertion:AuthenticatingAuthority');
+
+        return new self(
+            empty($authnContextClassRef) ? null : AuthnContextClassRef::fromXML($authnContextClassRef[0]),
+            empty($authnContextDecl) ? null : AuthnContextDecl::fromXML($authnContextDecl[0]),
+            empty($authnContextDeclRef) ? null : AuthnContextDeclRef::fromXML($authnContextDeclRef[0]),
+            array_map([AuthenticatingAuthority::class, 'fromXML'], $authenticatingAuthorities) ?: null
+        );
+    }
+
+
+    /**
+     * Convert this AuthContextDeclRef to XML.
+     *
+     * @param \DOMElement|null $parent The element we should append this AuthnContextDeclRef to.
+     * @return \DOMElement
+     */
+    public function toXML(DOMElement $parent = null): DOMElement
+    {
+        if ($parent === null) {
+            $doc = DOMDocumentFactory::create();
+            $e = $doc->createElementNS(Constants::NS_SAML, 'saml:AuthnContext');
+            $doc->appendChild($e);
+        } else {
+            $e = $parent->ownerDocument->createElementNS(Constants::NS_SAML, 'saml:AuthnContext');
+            $parent->appendChild($e);
+        }
+
+        if (!empty($this->authnContextClassRef)) {
+            $this->authnContextClassRef->toXML($e);
+
+            Assert::oneOf(
+                null,
+                [$this->authnContextDecl, $this->authnContextDeclRef],
+                'Can only have one of AuthnContextDecl/AuthnContextDeclRef'
+            );
+        } else {
+            Assert::false(
+                empty($this->authnContextDecl) && empty($this->authnContextDeclRef),
+                'You need either an AuthnContextDecl or an AuthnContextDeclRef'
+            );
+        }
```

Comment:
If the setters are protected (or private), then every possible path to create an object will lead to this check in the constructor. Then the check becomes redundant here.